File: bg/bobsgameonlinejava/src/main/java/com/bobsgame/client/engine/EnginePart.java
package com.bobsgame.client.engine;

import org.slf4j.LoggerFactory;

import ch.qos.logback.classic.Logger;

import com.bobsgame.ClientMain;
import com.bobsgame.client.ControlsManager;
import com.bobsgame.client.engine.cinematics.CinematicsManager;
import com.bobsgame.client.engine.entity.Cameraman;
import com.bobsgame.client.engine.entity.SpriteManager;
import com.bobsgame.client.engine.event.ActionManager;
import com.bobsgame.client.engine.event.EventManager;
import com.bobsgame.client.engine.game.Clock;
import com.bobsgame.client.engine.game.FriendManager;
import com.bobsgame.client.engine.game.ClientGameEngine;
import com.bobsgame.client.engine.game.Player;
import com.bobsgame.client.engine.game.Wallet;
import com.bobsgame.client.engine.game.gui.*;
import com.bobsgame.client.engine.game.gui.gameStore.GameStore;
import com.bobsgame.client.engine.game.gui.statusbar.*;
import com.bobsgame.client.engine.game.gui.stuffMenu.StuffMenu;
import com.bobsgame.client.engine.game.nd.*;
import com.bobsgame.client.engine.map.Map;
import com.bobsgame.client.engine.map.MapManager;
import com.bobsgame.client.engine.sound.AudioManager;
import com.bobsgame.client.engine.text.CaptionManager;
import com.bobsgame.client.engine.text.TextManager;
import com.bobsgame.client.network.GameClientTCP;
import com.bobsgame.net.GameSave;
import com.bobsgame.shared.MapData;
//=========================================================================================================================
public class EnginePart
{//=========================================================================================================================

	public static Logger log = (Logger) LoggerFactory.getLogger(EnginePart.class);


	private Engine e = null;

	private static ClientGameEngine clientGameEngine = null;
	private static ControlsManager controlsManager = null;


	private boolean isActivated = false;
	private long lastTimeHere = -1;
	private long timeActivated = -1;


	//=========================================================================================================================
	public EnginePart(Engine e)
	{//=========================================================================================================================
		this.e = e;

		//if(g==null)log.error("e is null");

	}

	public static void setClientGameEngine(ClientGameEngine gameEngine)
	{

		EnginePart.clientGameEngine = gameEngine;
	}

	public static void setControlsManager(ControlsManager controlsManager)
	{

		EnginePart.controlsManager = controlsManager;
	}






	public void toggleActivated()
	{
		isActivated = !isActivated;
		timeActivated = System.currentTimeMillis();


		//TODO: send flag update with this game object TYPEID

		//TODO: maybe instead of flag i should allow any kind of value: int, float, bool, string

		//so, GAMEVALUE
	}
	public void setActivated(boolean b)
	{
		isActivated = b;
		timeActivated = System.currentTimeMillis();

		//TODO: send flag update with this game object TYPEID
	}
	public boolean isActivated()
	{
		return isActivated;
	}
	public boolean wasEverActivated()
	{
		if(timeActivated==-1)return false;
		return true;
	}
	public int secondsSinceActivated()
	{
		if(timeActivated==-1)return 0;

		return (int)((System.currentTimeMillis()-timeActivated)/1000);
	}
	public boolean wasEverHere()
	{
		if(lastTimeHere==-1)return false;
		return true;
	}
	public void resetLastTimeHere()
	{
		lastTimeHere = -1;
	}
	public void setLastTimeHere()
	{
		lastTimeHere = System.currentTimeMillis();

		//TODO: send flag update with this game object TYPEID
	}
	public long getLastTimeHere()
	{
		return lastTimeHere;
	}
	public int secondsSinceLastHere()
	{
		if(lastTimeHere==-1)return -1;
		return (int)((System.currentTimeMillis()-lastTimeHere)/1000);
	}
	public int minutesSinceLastHere()
	{
		if(lastTimeHere==-1)return -1;
		return (int)(((System.currentTimeMillis()-lastTimeHere)/1000)/60);
	}




	public Engine Engine()
	{
		return e;
	}
	public Cameraman Cameraman()
	{
		return e.cameraman;
	}
	public MapManager MapManager()
	{
		return e.mapManager;
	}
	public SpriteManager SpriteManager()
	{
		return e.spriteManager;
	}
	public ActionManager ActionManager()
	{
		return e.actionManager;
	}
	public TextManager TextManager()
	{
		return e.textManager;
	}
	public AudioManager AudioManager()
	{
		return e.audioManager;
	}
	public CaptionManager CaptionManager()
	{
		return e.captionManager;
	}
	public EventManager EventManager()
	{
		return e.eventManager;
	}
	public CinematicsManager CinematicsManager()
	{
		return e.cinematicsManager;
	}

	public Map CurrentMap()
	{
		Map m = e.mapManager.currentMap;
		if(m==null)m = new Map(e,new MapData(-1,"none",0,0));
		return m;
	}


















	public static ControlsManager ControlsManager()
	{
		return controlsManager;
	}
	public static ClientGameEngine ClientGameEngine()
	{
		return clientGameEngine;
	}

	public static GameClientTCP Network()
	{
		if(ClientMain.clientMain==null)return null;
		return ClientMain.clientMain.gameClientTCP;
	}

	public static Clock Clock()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().clock;
	}
	public static GUIManager GUIManager()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().guiManager;
	}
	public static StuffMenu StuffMenu()
	{
		if(ClientGameEngine()==null)return null;
		return GUIManager().stuffMenu;
	}
	public static GameStore GameStore()
	{
		if(ClientGameEngine()==null)return null;
		return GUIManager().gameStore;
	}
	public static PlayerEditMenu PlayerEditMenu()
	{
		if(ClientGameEngine()==null)return null;
		return GUIManager().playerEditMenu;
	}
	public static Player Player()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().player;
	}
	public static ND ND()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().nD;
	}
	public static Wallet Wallet()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().wallet;
	}
	public static FriendManager FriendManager()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().friendManager;
	}
	public static StatusBar StatusBar()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().statusBar;
	}
	public static NotificationManager NotificationManager()
	{
		if(ClientGameEngine()==null)return null;
		return StatusBar().notificationManager;
	}
	synchronized public static GameSave GameSave()
	{
		if(ClientGameEngine()==null)return null;
		return ClientGameEngine().gameSave_S();
	}



//
//	public String getIDString()
//	{
//
//		return null;
//	}
//
//
//
//	public String getShortTypeName()
//	{
//
//		return null;
//	}




}

File: bg/okgame/legacy-src/src/Engine/EnginePart.cpp
#include "stdafx.h"

//------------------------------------------------------------------------------
//Copyright Robert Pelloni.
//All Rights Reserved.
//------------------------------------------------------------------------------

//#pragma once



Logger EnginePart::log = Logger("EnginePart");

//BGClientEngine* EnginePart::clientGameEngine = nullptr;
//ControlsManager* EnginePart::controlsManager = nullptr;

EnginePart::EnginePart()
{ //=========================================================================================================================
}

EnginePart::EnginePart(Engine* g)
{ //=========================================================================================================================
	this->e = g;

	//if(g==null)log.error("e is null");
}

//
//void EnginePart::setControlsManager(ControlsManager* controlsManager)
//{
//	EnginePart::controlsManager = controlsManager;
//}

void EnginePart::toggleActivated()
{
	isActivated = !isActivated;
	timeActivated = System::currentHighResTimer();

	//TODO: send flag update with this game object TYPEID

	//TODO: maybe instead of flag i should allow any kind of value: int, float, bool, string

	//so, GAMEVALUE
}

void EnginePart::setActivated(bool b)
{
	isActivated = b;
	timeActivated = System::currentHighResTimer();

	//TODO: send flag update with this game object TYPEID
}

bool EnginePart::getIsActivated()
{
	return isActivated;
}

bool EnginePart::getWasEverActivated()
{
	if (timeActivated == -1)
	{
		return false;
	}
	return true;
}

int EnginePart::getSecondsSinceActivated()
{
	if (timeActivated == -1)
	{
		return 0;
	}

	long long startTime = timeActivated;
	long long currentTime = System::currentHighResTimer();
	int ticksPassed = (int)(System::getTicksBetweenTimes(startTime, currentTime));

	return (int)(ticksPassed / 1000);
}

bool EnginePart::getWasEverHere()
{
	if (lastTimeHere == -1)
	{
		return false;
	}
	return true;
}

void EnginePart::resetLastTimeHere()
{
	lastTimeHere = -1;
}

void EnginePart::setLastTimeHere()
{
	lastTimeHere = System::currentHighResTimer();

	//TODO: send flag update with this game object TYPEID
}

long long EnginePart::getLastTimeHere()
{
	return lastTimeHere;
}

int EnginePart::getSecondsSinceLastHere()
{
	if (lastTimeHere == -1)
	{
		return -1;
	}

	long long startTime = lastTimeHere;
	long long currentTime = System::currentHighResTimer();
	int ticksPassed = (int)(System::getTicksBetweenTimes(startTime, currentTime));

	return (int)(ticksPassed / 1000);
}

int EnginePart::getMinutesSinceLastHere()
{
	if (lastTimeHere == -1)
	{
		return -1;
	}

	long long startTime = lastTimeHere;
	long long currentTime = System::currentHighResTimer();
	int ticksPassed = (int)(System::getTicksBetweenTimes(startTime, currentTime));

	return (int)((ticksPassed / 1000) / 60);
}

Engine* EnginePart::getEngine()
{
	if(e==nullptr)return Main::gameEngine;
	return e;
}

shared_ptr<Cameraman> EnginePart::getCameraman()
{
	return getEngine()->cameraman;
}

shared_ptr<MapManager> EnginePart::getMapManager()
{
	return getEngine()->mapManager;
}

shared_ptr<SpriteManager> EnginePart::getSpriteManager()
{
	return getEngine()->spriteManager;
}

shared_ptr<ActionManager> EnginePart::getActionManager()
{
	return getEngine()->actionManager;
}

shared_ptr<TextManager> EnginePart::getTextManager()
{
	return getEngine()->textManager;
}

shared_ptr<AudioManager> EnginePart::getAudioManager()
{
	return getEngine()->audioManager;
}

shared_ptr<CaptionManager> EnginePart::getCaptionManager()
{
	return getEngine()->captionManager;
}

shared_ptr<EventManager> EnginePart::getEventManager()
{
	return getEngine()->eventManager;
}

shared_ptr<CinematicsManager> EnginePart::getCinematicsManager()
{
	return getEngine()->cinematicsManager;
}

shared_ptr<Map> EnginePart::getCurrentMap()
{
	return getEngine()->mapManager->currentMap;
}




//void EnginePart::setClientGameEngine(BGClientEngine* gameEngine)
//{
//	EnginePart::clientGameEngine = gameEngine;
//}

BGClientEngine* EnginePart::getClientGameEngine()
{
	return Main::gameEngine;
}

shared_ptr<ControlsManager> EnginePart::getControlsManager()
{
	return getEngine()->getControlsManager();
}


Clock* EnginePart::getClock()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->clock;
}

GUIManager* EnginePart::getGUIManager()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->guiManager;
}

StuffMenu* EnginePart::getStuffMenu()
{
	if (getGUIManager() == nullptr)return nullptr;
	return getGUIManager()->stuffMenu;
}

GameStore* EnginePart::getGameStore()
{
	if (getGUIManager() == nullptr)return nullptr;
	return getGUIManager()->gameStore;
}

PlayerEditMenu* EnginePart::getPlayerEditMenu()
{
	if (getGUIManager() == nullptr)return nullptr;
	return getGUIManager()->playerEditMenu;
}

Player* EnginePart::getPlayer()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->player;
}

ND* EnginePart::getND()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->nD;
}

Wallet* EnginePart::getWallet()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->wallet;
}

FriendManager* EnginePart::getFriendManager()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->friendManager;
}

BobStatusBar* EnginePart::getBobStatusBar()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getClientGameEngine()->statusBar;
}

NotificationManager* EnginePart::getNotificationManager()
{
	if (getClientGameEngine() == nullptr)return nullptr;
	return getBobStatusBar()->notificationManager;
}

TCPServerConnection* EnginePart::getServerConnection()
{
	return &BobNet::tcpServerConnection;
}

GameSave EnginePart::getGameSave()
{
	return getServerConnection()->getGameSave_S();
}

//=========================================================================================================================
bool EnginePart::udpPeerMessageReceived(UDPPeerConnection* c, string e)
{//=========================================================================================================================

	string s = e;

	if (s.find(":") == string::npos)return false;

	string command = s.substr(s.find(":") + 1);
	s = s.substr(s.find(":") + 1);

	return false;
}